### ISSUE TITLE
Use consistent (Scala-specific) (Java-specific) qualifiers in method scaladoc

### ISSUE THREAD

**Issue by heuermh** [bigdatagenomics/adam#1986]
[Apache Spark scaladoc](https://spark.apache.org/docs/2.1.0/api/java/org/apache/spark/sql/Dataset.html#filter(java.lang.String)) uses (Scala-specific) (Java-specific) qualifiers on methods, e.g.
(Scala-specific) Returns a new Dataset that contains the result of applying func to each element.
(Java-specific) Returns a new Dataset that contains the result of applying func to each element.

This is preferable to a mention in second or third sentences as we often do, since those don't show up in the scaladoc method overview.

**Comment by heuermh** [comment]
@devin-petersohn @fnothaft While going through this, I noticed that we have `flankSize` as `java.lang.Integer`, `java.lang.Double`, and `Long` in separate region join methods.

Is there any reason not to have just (Java-specific) `java.lang.Long` and (Scala-specific) `Long` methods?  E.g.
```scala
  // (Java-specific)
  def broadcastRegionJoin[X, Y <: Product, Z <: GenomicDataset[X, Y, Z]](
    genomicDataset: GenomicDataset[X, Y, Z],
    flankSize: java.lang.Long): GenericGenomicDataset[(T, X), (U, Y)] = {

    implicit val tTag = ClassTag.AnyRef.asInstanceOf[ClassTag[T]]
    implicit val xTag = ClassTag.AnyRef.asInstanceOf[ClassTag[X]]
    implicit val txTag = ClassTag.AnyRef.asInstanceOf[ClassTag[(T, X)]]
    implicit val u1Tag: TypeTag[U] = uTag
    implicit val u2Tag: TypeTag[Y] = genomicDataset.uTag
    implicit val uyTag = typeTag[(U, Y)]

    broadcastRegionJoin(genomicDataset, flankSize)
  }

  // (Scala-specific)
  def broadcastRegionJoin[X, Y <: Product, Z <: GenomicDataset[X, Y, Z]](
    genomicDataset: GenomicDataset[X, Y, Z],
    flankSize: Long)(
      implicit tTag: ClassTag[T],
      xTag: ClassTag[X],
      txTag: ClassTag[(T, X)],
      uyTag: TypeTag[(U, Y)]): GenericGenomicDataset[(T, X), (U, Y)] = InnerBroadcastJoin.time {
    // ...
  }
```

**Comment by devin-petersohn** [comment]
That's fine with me. I haven't been in the code for over a year, and the original `flanksize` was `Long`, but it looks like some more recent commits have changed that. I will defer to @fnothaft and others currently maintaining.

**Comment by heuermh** [comment]
Thanks, @devin-petersohn.  Perhaps there is an R or Python reason for this; I haven't seen it yet.

**Comment by heuermh** [comment]
Yeah, appears that calling from Python causes trouble, see Jenkins failures e.g. https://amplab.cs.berkeley.edu/jenkins//job/ADAM-prb/2894/
```
bdgenomics/adam/test/alignmentRecordDataset_test.py:359: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
bdgenomics/adam/rdd.py:541: in rightOuterShuffleRegionJoinAndGroupByLeft
    return GenomicDataset(self._jvmRdd.rightOuterShuffleRegionJoinAndGroupByLeft(genomicRdd._jvmRdd, flankSize),
/tmp/adamTestGtuGXy5/deleteMePleaseThisIsNoLongerNeeded/spark-2.2.2-bin-hadoop2.7/python/lib/py4j-0.10.7-src.zip/py4j/java_gateway.py:1257: in __call__
    answer, self.gateway_client, self.target_id, self.name)
../spark-2.2.2-bin-hadoop2.7/python/pyspark/sql/utils.py:63: in deco
    return f(*a, **kw)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

answer = 'xro7372'
gateway_client = <py4j.java_gateway.GatewayClient object at 0x7fc16ab58fd0>
target_id = 'o7369', name = 'rightOuterShuffleRegionJoinAndGroupByLeft'

    def get_return_value(answer, gateway_client, target_id=None, name=None):
        """Converts an answer received from the Java gateway into a Python object.
    
        For example, string representation of integers are converted to Python
        integer, string representation of objects are converted to JavaObject
        instances, etc.
    
        :param answer: the string returned by the Java gateway
        :param gateway_client: the gateway client used to communicate with the Java
            Gateway. Only necessary if the answer is a reference (e.g., object,
            list, map)
        :param target_id: the name of the object from which the answer comes from
            (e.g., *object1* in `object1.hello()`). Optional.
        :param name: the name of the member from which the answer comes from
            (e.g., *hello* in `object1.hello()`). Optional.
        """
        if is_error(answer)[0]:
            if len(answer) > 1:
                type = answer[1]
                value = OUTPUT_CONVERTER[type](answer[2:], gateway_client)
                if answer[1] == REFERENCE_TYPE:
                    raise Py4JJavaError(
                        "An error occurred while calling {0}{1}{2}.\n".
>                       format(target_id, ".", name), value)
E                   Py4JJavaError: An error occurred while calling o7369.rightOuterShuffleRegionJoinAndGroupByLeft.
E                   : java.lang.StackOverflowError
E                   	at org.bdgenomics.adam.rdd.GenomicDataset$class.rightOuterShuffleRegionJoinAndGroupByLeft(GenomicDataset.scala:2551)
E                   	at org.bdgenomics.adam.rdd.AvroGenomicDataset.rightOuterShuffleRegionJoinAndGroupByLeft(GenomicDataset.scala:3536)
E                   	at org.bdgenomics.adam.rdd.GenomicDataset$class.rightOuterShuffleRegionJoinAndGroupByLeft(GenomicDataset.scala:2555)
E                   	at org.bdgenomics.adam.rdd.AvroGenomicDataset.rightOuterShuffleRegionJoinAndGroupByLeft(GenomicDataset.scala:3536)
E                   	at org.bdgenomics.adam.rdd.GenomicDataset$class.rightOuterShuffleRegionJoinAndGroupByLeft(GenomicDataset.scala:2555)
E                   	at org.bdgenomics.adam.rdd.AvroGenomicDataset.rightOuterShuffleRegionJoinAndGroupByLeft(GenomicDataset.scala:3536)
```

**Comment by heuermh** [comment]
@fnothaft Join methods with `java.lang.Double` and `java.lang.Integer` were added in commit 
https://github.com/bigdatagenomics/adam/commit/a28da442150401de6560d589402f152f53729942

I recall some frustration with calling methods from Python, could you summarize that again?  I've probably jumped back into that mess with changes proposed in #2097.
